scipy/io/matlab/mio.py

```diff
@@ -14,30 +14,40 @@
 __all__ = ['mat_reader_factory', 'loadmat', 'savemat', 'whosmat']
 
 
-def _open_file(file_like, appendmat):
+@contextmanager
+def _open_file_context(file_like, appendmat, mode='rb'):
+    f, opened = _open_file(file_like, appendmat, mode)
+    yield f
+    if opened:
+        f.close()
+
+
+def _open_file(file_like, appendmat, mode='rb'):
     """
     Open `file_like` and return as file-like object. First, check if object is
     already file-like; if so, return it as-is. Otherwise, try to pass it
     to open(). If that fails, and `file_like` is a string, and `appendmat` is true,
     append '.mat' and try again.
     """
-    try:
-        file_like.read(0)
+    if hasattr(file_like, 'read') and hasattr(file_like, 'write'):
```

Comment:
Can we get away with only checking 'write' if `mode` says it's necessary?

Reply:
I don't think that this is going to be a performance hit.

Reply:
Even a non-writeable file-like should still have a write method.

Reply:
Sure, generally it will, it's just a theoretical wart, if I can put it that way.